### PR TITLE
New: ConnectAd PreBid Server Adapter

### PR DESCRIPTION
**Pull request by rtuschkany**
New PreBid Server Adapter for ConnectAd

### CODE REVIEW

**Comment by SyntaxNode** [comment]
Thank you for your contribution. We'll review this PR early next week.

For now, could you please open a PR in https://github.com/prebid/prebid.github.io to add a doc for this new adapter. More information is found in our new adapter guide: https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#document-your-adapter

**Comment by rtuschkany** [comment]
I just opened a PR to update our adapter docs: https://github.com/prebid/prebid.github.io/pull/2305

File: adapters/connectad/connectad.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+type ConnectAdAdapter struct {
+	endpoint string
+}
+
+type connectadImpExt struct {
+	ConnectAd openrtb_ext.ExtImpConnectAd `json:"connectad"`
+}
+
+func (a *ConnectAdAdapter) SkipNoCookies() bool {
```
Comment (SyntaxNode) [inline adapters/connectad/connectad.go:24]: `SkipNoCookies` is part of the legacy interface. It is no longer used. Please remove it.

File: adapters/connectad/connectad.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	return &ConnectAdAdapter{
+		endpoint: endpoint,
+	}
+}
+
+func (a *ConnectAdAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+
+	var errs []error
+
+	if len(request.Imp) == 0 {
```
Comment (SyntaxNode) [inline adapters/connectad/connectad.go:38]: You can leave this check in here if you wish, other adapters do.. but it shouldn't be necessary. The core code will not call `MakeRequests` with no impressions.

File: adapters/connectad/connectad.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("No Imps in Bid Request"),
+		}}
+	}
+
+	if errs := preprocess(request); errs != nil && len(errs) > 0 {
```
Comment (SyntaxNode) [inline adapters/connectad/connectad.go:44]: The `errs != nil` isn't necessary. Go's `len` gracefully handles nil.

File: adapters/connectad/connectad.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Error in packaging request to JSON"),
+		}}
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	addHeaderIfNonEmpty(headers, "User-Agent", request.Device.UA)
+	addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IP)
```
Comment (SyntaxNode) [inline adapters/connectad/connectad.go:61]: You might want to consider that request.Device.IPv6 might be populated instead.

File: adapters/connectad/connectad.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+	if len(headerValue) > 0 {
+		headers.Add(headerName, headerValue)
+	}
+}
+
+func unpackImpExt(imp *openrtb.Imp) (*openrtb_ext.ExtImpConnectAd, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: err.Error(),
```
Comment (SyntaxNode) [inline adapters/connectad/connectad.go:167]: Consider including the impression index as you do with the other unmarshal failure branches.

File: adapters/connectad/connectad.go [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+	if err != nil {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Error in packaging request to JSON"),
+		}}
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	addHeaderIfNonEmpty(headers, "User-Agent", request.Device.UA)
```
Comment (SyntaxNode) [inline adapters/connectad/connectad.go:60]: Be careful. request.Device can be nil here. Please wrap with a `requestDevice != nil` check.

File: adapters/connectad/usersync_test.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+	syncURL := "https://cdn.connectad.io/connectmyusers.php?gdpr={{.GDPR}}&consent={{.GDPRConsent}}&us_privacy={{.USPrivacy}}&cb=localhost%2Fsetuid%3Fbidder%3Dconnectad%26gdpr%3D{{.GDPR}}%26gdpr_consent%3D{{.GDPRConsent}}%26uid%3D%24UID"
+	syncURLTemplate := template.Must(
+		template.New("sync-template").Parse(syncURL),
+	)
+
+	syncer := NewConnectAdSyncer(syncURLTemplate)
+	syncInfo, err := syncer.GetUsersyncInfo(privacy.Policies{
+		GDPR: gdpr.Policy{
+			Signal:  "1",
+			Consent: "BOPVK28OVJoTBABABAENBs-AAAAhuAKAANAAoACwAGgAPAAxAB0AHgAQAAiABOADkA",
```
Comment (SyntaxNode) [inline adapters/connectad/usersync_test.go:23]: Consider using an obviously fake consent string here. This is only testing string substitution.

File: openrtb_ext/imp_connectad.go [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package openrtb_ext
+
+type ExtImpConnectAd struct {
+	NetworkId int     `json:"networkId"`
```
Comment (SyntaxNode) [inline openrtb_ext/imp_connectad.go:4]: Per Go naming convention, abbreviations and acronyms should be upper case.
NetworkId -> NetworkID
SiteId -> SiteID

**Review by SyntaxNode — COMMENTED** [review]
Thank you for your contribution. The unit test coverage rate is 81% and we strive for higher coverage. I understand that Marshal errors can't be effectively tested, but can you please increase test coverage for other failure branches? such as with:
- A `preprocess` error.
- `request.Device.DNT == nil`
- `httpRes.StatusCode != http.StatusOK` 
- `json.Unmarshal(httpRes.Body, &bidResp)` 
- `unpackImpExt` failure branches
- `imp.Banner == nil`
- `len(banner.Format) == 0`

File: static/bidder-params/connectad.json [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    "properties": {
+      "networkId": {
+        "type": "integer",
+        "description": "NetworkId"
+      },
+      "siteId": {
+        "type": "integer",
+        "description": "SiteId"
+      },
+      "bidfloor": {
```
Comment (SyntaxNode) [inline static/bidder-params/connectad.json:17]: I noticed the doc PR has this listed as `floorprice`.

Comment (rtuschkany) [reply static/bidder-params/connectad.json:17]: oh boy, so true! we are accepting both values in the backend. I will update the docs accordingly

File: adapters/connectad/connectad.go [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+		endpoint: endpoint,
+	}
+}
+
+func (a *ConnectAdAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+
+	var errs []error
+
+	if errs := preprocess(request); len(errs) > 0 {
+		return nil, append(errs, &errortypes.BadInput{
```
Comment (SyntaxNode) [inline adapters/connectad/connectad.go:35]: The error message is being appended to the list of errors, making the including of the `errs` as a string redundant. Please consider passing the errors directly through or removing the `errs` part of the general message if you still want to provide clear direction for where the errors are coming from.

File: adapters/connectad/connectad.go [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+		addHeaderIfNonEmpty(headers, "Accept-Language", request.Device.Language)
+		if request.Device.IP != "" {
+			addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IP)
+		} else if request.Device.IPv6 != "" {
+			addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IPv6)
+		}
+		if request.Device.DNT != nil {
+			addHeaderIfNonEmpty(headers, "DNT", strconv.Itoa(int(*request.Device.DNT)))
+		} else {
+			addHeaderIfNonEmpty(headers, "DNT", "0")
```
Comment (SyntaxNode) [inline adapters/connectad/connectad.go:61]: Please add a test case for `DNT == nil`.

File: adapters/connectad/connectad.go [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+
+	var bidResp openrtb.BidResponse
+
+	if err := json.Unmarshal(httpRes.Body, &bidResp); err != nil {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unable to unpackage bid response. Error: %s", err.Error()),
+		}}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
```
Comment (SyntaxNode) [inline adapters/connectad/connectad.go:93]: Please consider setting capacity to `len(bidResp.SeatBid)` to slightly improve memory allocation.

File: adapters/connectad/connectad.go (hunk 5 above)
Comment (SyntaxNode) [reply adapters/connectad/connectad.go:163]: The code didn't change in this area. Do you feel that including the imp id here is not useful?

File: adapters/connectad/connectad.go [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+		}
+	}
+
+	return &cadExt, nil
+}
+
+func buildImpBanner(imp *openrtb.Imp) error {
+	imp.Ext = nil
+
+	if imp.Banner == nil {
```
Comment (SyntaxNode) [inline adapters/connectad/connectad.go:186]: Please add a unit test for this branch.

File: adapters/connectad/connectad.go [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("We need a Banner Object in the request"),
+		}
+	}
+
+	bannerCopy := *imp.Banner
```
Comment (SyntaxNode) [inline adapters/connectad/connectad.go:192]: Consider moving the creation of the copy within the `banner.W == nil && banner.H == nil` block to avoid the memory allocation if it's not needed.

File: adapters/connectad/usersync_test.go [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+	syncURL := "https://cdn.connectad.io/connectmyusers.php?gdpr={{.GDPR}}&consent={{.GDPRConsent}}&us_privacy={{.USPrivacy}}&cb=localhost%2Fsetuid%3Fbidder%3Dconnectad%26gdpr%3D{{.GDPR}}%26gdpr_consent%3D{{.GDPRConsent}}%26uid%3D%24UID"
+	syncURLTemplate := template.Must(
+		template.New("sync-template").Parse(syncURL),
+	)
+
+	syncer := NewConnectAdSyncer(syncURLTemplate)
+	syncInfo, err := syncer.GetUsersyncInfo(privacy.Policies{
+		GDPR: gdpr.Policy{
+			Signal:  "1",
+			Consent: "BOPVAAoACwAGgAPAAxAB0ADkA",
```
Comment (SyntaxNode) [inline adapters/connectad/usersync_test.go:23]: Please consider using obviously fake values here (and for CCPA), such as "A" and "B" or some such to clarify that we're only resting string substitution.

**Review by SyntaxNode — COMMENTED** [review]
Thank you very much for a quick update. I have just a few more comments.

**Comment by rtuschkany** [comment]
> Thank you very much for a quick update. I have just a few more comments.

Thanks for the very quick review. Just pushed an update

**Review by SyntaxNode — APPROVED** [review]
Looks great. Thank you. @bsardo will look at this soon for a second approval. Until then, looks like you have a conflict to resolve.

**Comment by bsardo** [comment]
Looks good @rtuschkany. Just a couple of questions and suggestions.

File: adapters/connectad/connectadtest/supplemental/nobanner.json [hunk 16]
```diff
@@ -0,0 +1,35 @@
+{
+    "mockBidRequest": {
+      "id": "test-request-id",
+      "imp": [
+        {
+          "id": "test-banner-imp-id",
+          "banner": {
+            "format": [
+            ]
+          },
```
[the remark this answers is not in the document]
Comment (rtuschkany) [reply adapters/connectad/connectadtest/supplemental/nobanner.json]: @bsardo thx!! file was replaced by no_banner.json to have a consistent naming

File: openrtb_ext/imp_connectad.go [hunk 17]
```diff
@@ -0,0 +1,7 @@
+package openrtb_ext
+
+type ExtImpConnectAd struct {
+	NetworkID int     `json:"networkId"`
```
[the remark this answers is not in the document]
Comment (rtuschkany) [reply openrtb_ext/imp_connectad.go:4]: @bsardo that's intentional and ok for now.

**Comment by bsardo** [comment]
Looks good @rtuschkany, just need to resolve the conflict.

**Comment by SyntaxNode** [comment]
@rtuschkany This looks good to us. We need you to resolve the conflicts before we can merge it in.

**Comment by rtuschkany** [comment]
> @rtuschkany This looks good to us. We need you to resolve the conflicts before we can merge it in.

Done!

**Comment by SyntaxNode** [comment]
Huh. Strange. I still see the "This branch has conflicts that must be resolved" warning on this PR. Did you merge in master for this branch? See https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork

**Comment by rtuschkany** [comment]
@SyntaxNode I had some errors trying to sync. Just opened a new PR (https://github.com/prebid/prebid-server/pull/1505) from a fresh fork. Please close this PR. Thanks!
